frontends/concrete-python/concrete/fhe/mlir/converter.py

```diff
@@ -814,4 +822,46 @@ def zeros(self, ctx: Context, node: Node, preds: List[Conversion]) -> Conversion
         assert len(preds) == 0
         return ctx.zeros(ctx.typeof(node))
 
+    def tfhers_to_native(self, ctx: Context, node: Node, preds: List[Conversion]) -> Conversion:
+        assert len(preds) == 1
+        tfhers_int = preds[0]
+        result_bit_width, pad_width, msg_width = (
+            node.properties["attributes"]["result_bit_width"],
+            node.properties["attributes"]["pad_width"],
+            node.properties["attributes"]["msg_width"],
+        )
+
+        # result will remove the last dim which is the dim of ciphertexts
+        result_shape = tfhers_int.shape[:-1]
+        # if the result shape is () then the types would actually be scalars
+        result_type = ctx.tensor(ctx.eint(result_bit_width), result_shape)
+        extract_type = ctx.tensor(ctx.eint(msg_width), result_shape)
+
+        n_elem = tfhers_int.shape[-1]
+        acc = ctx.zeros(result_type)
+        # index should span all dims except the last one which select a single elem
+        index = [slice(None, None, None) for _ in range(len(tfhers_int.shape) - 1)]
+        index.append(0)
+        # start with lsb
+        for i in range(n_elem - 1, -1, -1):
+            index[-1] = i
+            ct = ctx.index_static(extract_type, tfhers_int, index)
+            # considering we are iterating the ciphertexts lsb first, we will
+            # shift `msg_width` bits to the left starting from 0
+            shift = msg_width * (n_elem - 1 - i)
+            tlu = [v * 2**shift for v in range(2**msg_width)]
+
+            to_add = ctx.tlu(result_type, ct, tlu)
+            acc = ctx.add(result_type, acc, to_add)
+        return acc
```

Comment:
Please don't do it like this. Have a multi tlu on the whole `tfhers_int`, and then a `sum` on it.

Here is the idea:
```python
        tables = [
            [value << (msg_width * (n_elem - 1 - i)) for value in range(2**msg_width)]
            for i in range(n_elem)
        ]
        mapping = np.broadcast_to(np.array(list(reversed(range(n_elem)))), tfhers_int.shape)
        mapped = ctx.tlu(result_type, tfhers_int, tables, mapping)
        result = ctx.sum(result_type, tfhers_int, axes=-1)
                       # ^^^^^^^^^^^ types might be slightly different
```
This code can probably be improved, but it shows the idea!

Reply:
I wanted to have a first prototype, and the goal was indeed to make it via tensor operations, so thanks for providing a code snippet for improving it